utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,567 @@
+  setsMap :: Map Int (Vector DescriptorSetLayoutBinding)
+  setsMap =
+    M.fromList
+      . map (liftA2 (,) (fst . head) (V.fromList . (snd <$>)))
+      . (NE.toList <$>)
+      . NE.groupAllWith fst
+      . V.toList
+      $ bindings
+
+makeDescriptorSetLayoutCreateInfo
+  :: Vector DescriptorSetLayoutBinding -> DescriptorSetLayoutCreateInfo '[]
+makeDescriptorSetLayoutCreateInfo bindings = zero { bindings = bindings }
+
+data VertexAttributeType
+  = Vec2
+  | Vec3
+  | Vec4
+  deriving (Show)
+
+vertexAttributeTypeFromText :: Text -> Maybe VertexAttributeType
+vertexAttributeTypeFromText = \case
+  "vec2" -> Just Vec2
+  "vec3" -> Just Vec3
+  "vec4" -> Just Vec4
+  _      -> Nothing
+
+vertexAttributeTypeToText :: VertexAttributeType -> Text
+vertexAttributeTypeToText = \case
+  Vec2 -> "vec2"
+  Vec3 -> "vec3"
+  Vec4 -> "vec4"
+
+instance FromJSON VertexAttributeType where
+  parseJSON = withTextMaybe "type" vertexAttributeTypeFromText
+
+instance ToJSON VertexAttributeType where
+  toJSON = String . vertexAttributeTypeToText
+
+convertVertexAttributeType :: VertexAttributeType -> (Word32, Format)
+convertVertexAttributeType = \case
+  Vec2 -> (2 * floatSize, FORMAT_R32G32_SFLOAT)
+  Vec3 -> (3 * floatSize, FORMAT_R32G32B32_SFLOAT)
+  Vec4 -> (4 * floatSize, FORMAT_R32G32B32A32_SFLOAT)
+  where floatSize = 4
+
+data VertexAttribute = VertexAttribute
+  { binding  :: Word32
+  , location :: Word32
+  , size     :: Word32
+  , format   :: Format
+  }
+  deriving Show
+
+-- only support single binding for input vertex
+-- [SaschaWillems - Multiple Vulkan buffer binding points](https://gist.github.com/SaschaWillems/428d15ed4b5d71ead462bc63adffa93a)
+-- maybe can provide a binding map parameter to specify individual binding by name, like `foo [(1, ["inPos", "inColor"]) (2, ["texCoord"])]` speficies that `(binding = 1) inPos, (binding = 1) inColor, (binding = 2) texCoord`
+-- [island pipeline](https://github.com/tgfrerer/island/blob/76d0d38cba74181fa3774cef38aba4d96b6861dc/modules/le_backend_vk/le_pipeline.cpp#L21)
+makePipelineVertexInputStateCreateInfo
+  :: Vector Input -> Maybe (PipelineVertexInputStateCreateInfo '[])
+makePipelineVertexInputStateCreateInfo []     = Nothing
+makePipelineVertexInputStateCreateInfo inputs = Just zero
+  { vertexBindingDescriptions
+  , vertexAttributeDescriptions
+  }
+ where
+  vertexAttributes =
+    join . V.map makeVertexAttribute $ inputs :: Vector VertexAttribute
+  vertexAttributeDescriptions =
+    makeVertexInputAttributeDescriptions vertexAttributes :: Vector
+        VertexInputAttributeDescription
+  vertexBindingDescriptions =
+    makeVertexInputBindingDescriptions vertexAttributes :: Vector
+        VertexInputBindingDescription
+
+makeVertexInputBindingDescriptions
+  :: Vector VertexAttribute -> Vector VertexInputBindingDescription
+makeVertexInputBindingDescriptions =
+  V.fromList
+    . map (makeVertexInputBindingDescription . calculate)
+    . (NE.toList <$>)
```

Comment:
Let the `calculate` enjoy the benefits of having an evidence of non-emptiness :smile:

Reply:
fixed.